### ISSUE TITLE
React Newsletter is changing state during rendering

### ISSUE THREAD

**Issue by peterbe** [mozilla/kuma#5782]
In https://sentry.prod.mozaws.net/operations/mdn-stage/issues/6247145/?query=is:unresolved we got an 500 error from `ssr-server.js`. But that's the underlying problem. The underlying problem was this:

```
Error: Minified React error #301; visit https://reactjs.org/docs/error-decoder.html?invariant=301 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Z (/app/kuma/javascript/dist/ssr.js:22:7658)
    at he (/app/kuma/javascript/dist/ssr.js:22:52879)
    at X (/app/kuma/javascript/dist/ssr.js:22:7139)
    at a (/app/kuma/javascript/dist/ssr.js:22:10878)
    at we (/app/kuma/javascript/dist/ssr.js:22:11874)
    at e.render (/app/kuma/javascript/dist/ssr.js:22:14460)
    at e.read (/app/kuma/javascript/dist/ssr.js:22:14084)
    at renderToString (/app/kuma/javascript/dist/ssr.js:22:19625)
    at gt (/app/kuma/javascript/dist/ssr.js:22:84247)
    at app.post (/app/kuma/javascript/ssr-server.js:68:14)
```

This happened because, during rendering, [state was changed](https://github.com/mozilla/kuma/blob/e1873a8914e1835e6c74774d6c764408fdf9e72d/kuma/javascript/src/newsletter.jsx#L159) which caused an infinite loop. 

The simplest solution is to fix `newsletter.jsx` to not change state during rendering but perhaps the best place to start is with the right tooling to prevent this from happening in the first place. 

Are there `eslint` plugins that catch state changes, in React hooks, during rendering? 

The error slipped through because both reviewer and test automation didn't catch it. 

I changed the code to:
```javascript
if (locale !== 'en-US') {
        throw new Error("BAL BLAB")
        // setShowNewsletterLang(true);
    }
```
and ran `npx jest` and it didn't fail. So, first of all we need much better test coverage on the `Newsletter` component. 

@Gregoor has [a PR](https://github.com/mozilla/kuma/pull/5772) that refactors `newsletter.jsx` but [it still has the bug](https://github.com/mozilla/kuma/blob/0533ef8d87acb3040e9221853c942820d41e59f6/kuma/javascript/src/newsletter.jsx#L154).

**Comment by peterbe** [comment]
Other things in `newsletter.jsx` that could do with some attention:

[This effect](https://github.com/mozilla/kuma/blob/e1873a8914e1835e6c74774d6c764408fdf9e72d/kuma/javascript/src/newsletter.jsx#L162-L168) [only fires on the first mount](https://twitter.com/ryanflorence/status/1125041041063665666) which means it could never work if the footer is re-used as you client-side navigation. 
Instead, `localStorage` should be used to exit early and return `null`. E.g. 

```javascript 
export default function Newsletter() {
    const someRef = useRef(...)
    useEffect(() => {
        ...
    });
    if (localStorage.getItem('newsletterHide') === 'true') {
        return null;
    }
    ...
}
```

**Comment by peterbe** [comment]
Why is it called `errorsArray`? Can't it be called `errors`? The plural (plus some Flow annotations) should make it clear that it's an array. In fact, if you make it like this instead:
```javascript
 const [errors, setErrors] = useState([]);
 ...

 return <div>
    {!!errors.length && <h4>There are problems!</h4>}
 </div>
```
then you don't even need to jump between `null`, an empty array, and a non-empty array.

**Comment by peterbe** [comment]
I don't even know what `skipFetch` is but [this doesn't feel right](https://github.com/mozilla/kuma/blob/e1873a8914e1835e6c74774d6c764408fdf9e72d/kuma/javascript/src/newsletter.jsx#L135-L137). Why can't that be state instead?

**Comment by schalkneethling** [comment]
> I don't even know what `skipFetch` is but [this doesn't feel right](https://github.com/mozilla/kuma/blob/e1873a8914e1835e6c74774d6c764408fdf9e72d/kuma/javascript/src/newsletter.jsx#L135-L137). Why can't that be state instead?

This is been like this historically:
https://github.com/mozilla/kuma/blob/master/kuma/static/js/newsletter.js#L41

The comments in the code suggest that there are instances when we could get back a response from the mozorg page that says something like:

```
{ success: false, errors: [] }
```

In this case, we do not know what went wrong and therefore do not have anything useful to show to the user. The decision was made to, in these cases, do a plain(non-XHR/non-Fetch) post to the `/newsletter` page on `mozorg` as it may be able to provide the actual error to the user.

**Comment by schalkneethling** [comment]
> [This effect](https://github.com/mozilla/kuma/blob/e1873a8914e1835e6c74774d6c764408fdf9e72d/kuma/javascript/src/newsletter.jsx#L162-L168) [only fires on the first mount](https://twitter.com/ryanflorence/status/1125041041063665666) which means it could never work if the footer is re-used as you client-side navigation.

True that. I was not thinking in terms of client-side navigation ;) This is worth doing though I reckon. @Gregoor, you want to work this and some of the other suggestions here into your refactor PR?

**Comment by Gregoor** [comment]
Thanks for the re-review Peter! That state setting in render I should've caught 🙈 Where is my mind?

Anyway, addressed them in the "refactor" PR. 

Though I didn't quite get the localStorage point. Afaik the only component that sets that is the Newsletter itself, no? If it is used by other component we might want to consider moving it into a context.

**Comment by peterbe** [comment]
I'm done butting into the PR work. I'm not sure if my contributions help or if it's just distracting. I knew we got ourselves into this mess and it's, as Schalk said, "going a million miles an hour" so we failed to test it properly both manually and automated. 
The release/launch is important but quality matters too. Let's all calm down and only do what we're comfortable with doing at a sane pace :)

(tried to find  a funny zen gif but couldn't find one so I'm not going to put one in)
